### PR TITLE
Add initial data for Deno

### LINKED ISSUE #5476: new engine/platform Deno
A relatively new platform Deno is under active development that includes language and Web API features. Deno seems well suited for addition as a target to track for compatibility in the same way Nodejs is currently tracked.

Please add Deno as an MDN target when it fits the requirements. I anticipate this platform displacing Nodejs due to overall design advantages.

More details:
https://github.com/denoland/deno
https://deno.land/

### PR DESCRIPTION
**Pull request by lucacasonato**
#### Summary

This PR adds BDC data for the Deno runtime. It adds versioned support data for all currently supported APIs and JS features, and negative support status to APIs which are either on the horizon for us, or likely for users to expect in Deno, or both.

#### Test results and supporting details

As per "Addition of browsers" section in the data guidelines, see the proposal here: https://github.com/mdn/browser-compat-data/issues/5476#issuecomment-854085351

Data is primarily based on my first hand knowledge of the platform, but all cross-referenced using the last 11 minor releases + our WPT data at https://wpt.deno.land. 

#### Related issues

Closes #5476

### CODE REVIEW

**Comment by foolip** [comment]
If you haven't seen it yet, you might want to check out https://mdn-bcd-collector.appspot.com/ which has generated tests for a lot of BCD. It would need some tweaking to work with Deno, but then you could verify a lot of this data, and most likely find many errors.

**Comment by lucacasonato** [comment]
@foolip Oh very nice. Is there a DOM free version of that? (JS only, with programmatic output?)

**Comment by foolip** [comment]
There isn't, but I'm thinking it shouldn't be a huge undertaking. All the tests are already in a generated tests.json and generating a big single script with all the feature detects from there shouldn't be too hard...

**Comment by lucacasonato** [comment]
WPT data is now also available on wpt.fyi: https://staging.wpt.fyi/results/?product=deno

File: api/CryptoKeyPair.json [hunk 1]
```diff
@@ -11,6 +11,9 @@
           "chrome_android": {
             "version_added": "37"
           },
+          "deno": {
+            "version_added": false
```
Comment (ddbeck) [inline api/CryptoKeyPair.json:15]: Some background which might be helpful for this and other Web APIs unsupported by Deno (e.g., `EventSource`):

On MDN we have some special handling for Web APIs for Node.js (and I expect, for Deno). Where there's no "nodejs" key, we don't show the column for Node; where there is, we show the column (even if it's `false` or `null`). This way, Node only appears on Web APIs that are relevant to non-browser usage.

(We don't do this for JavaScript features on Node. On JavaScript pages, we show that data is missing for Node.js.)

In other words, I'd only include Deno here if I wanted to highlight that it wasn't supported to developers using Deno. I suspect that's what you intended—given the support for other crypto features—but I thought I'd make it explicit.

File: api/Event.json [hunk 2]
```diff
@@ -1321,6 +1390,10 @@
               "version_added": "49",
               "notes": "Starting with Chrome 49, Firefox 54 and Opera 36, this property returns <a href='https://developer.mozilla.org/docs/Web/API/DOMHighResTimeStamp'><code>DOMHighResTimeStamp</code></a> instead of <a href='https://developer.mozilla.org/docs/Web/API/DOMTimeStamp'><code>DOMTimeStamp</code></a>."
             },
+            "deno": {
+              "version_added": "1.0",
+              "notes": "In Deno this represents the time since UNIX epoch, not since startup. See https://github.com/denoland/deno/issues/10835."
```
Comment (ddbeck) [inline api/Event.json:1395]:
```suggestion
              "notes": "In Deno this represents the time since UNIX epoch, not since startup. See <a href='https://github.com/denoland/deno/issues/10835'>#10835</a>."
```

File: api/ReadableByteStreamController.json [hunk 3]
```diff
@@ -11,6 +11,10 @@
           "chrome_android": {
             "version_added": "89"
           },
+          "deno": {
+            "version_added": "1.0",
+            "notes": "This interface is not exposed on the global scope on versions prior to 1.11."
```
Comment (ddbeck) [inline api/ReadableByteStreamController.json:16]: If instances are available, but `ReadableByteStreamController` isn't exposed, then I'd use two support statements:

1. Supported from version 1.11
2. version_added 1.0; version_removed 1.11 with `partial_implementation` and this note (without mentioning the version number)

File: api/ReadableStreamDefaultController.json [hunk 4]
```diff
@@ -11,6 +11,10 @@
           "chrome_android": {
             "version_added": "89"
           },
+          "deno": {
+            "version_added": "1.0",
+            "notes": "This interface was not exposed on the global scope until Deno 1.7."
```
Comment (ddbeck) [inline api/ReadableStreamDefaultController.json:16]: Another case where it'll be good to have two statements, one partial_implementation

File: api/ReadableStreamDefaultReader.json [hunk 5]
```diff
@@ -11,6 +11,10 @@
           "chrome_android": {
             "version_added": "78"
           },
+          "deno": {
+            "version_added": "1.0",
+            "notes": "This interface was not exposed on the global scope until Deno 1.7."
```
Comment (ddbeck) [inline api/ReadableStreamDefaultReader.json:16]: Another partial implementation

File: api/Request.json [hunk 6]
```diff
@@ -990,6 +1050,10 @@
             "chrome_android": {
               "version_added": "46"
             },
+            "deno": {
+              "version_added": "1.0",
+              "notes": "In Deno the 'manual' redirect mode does not result in a opaqueredirect response, but a regular response."
```
Comment (ddbeck) [inline api/Request.json:1055]: Some style suggestions for this note. Might not be a bad idea to include a similar note on the `Request` constructor, too, since that's where the redirect mode must be set

```suggestion
              "notes": "The <code>\"manual\"</code> redirect mode does not result in a <code>"opaqueredirect"</code> response, but a regular response."
```

File: api/TextDecoder.json [hunk 7]
```diff
@@ -150,6 +156,10 @@
             "chrome_android": {
               "version_added": "38"
             },
+            "deno": {
+              "version_added": "1.0",
+              "notes": "<code>stream</code> option not supported before Deno 1.11."
```
Comment (ddbeck) [inline api/TextDecoder.json:161]: The technically correct way to handle this is with a subfeature of `decode` documenting for `options.stream`. See https://github.com/mdn/browser-compat-data/blob/main/docs/data-guidelines.md#parameters-and-parameter-object-features

But since that would require touching some non-Deno data, I'm willing to overlook it here. 😃 I'd settle for a style change:

```suggestion
              "notes": "Before Deno 1.11, <code>option.stream</code> is not supported."
```

Better still would be a little more information on what "not supported" means here. Is the option ignored? Does it throw an error?

File: api/TransformStreamDefaultController.json [hunk 8]
```diff
@@ -10,6 +10,10 @@
           "chrome_android": {
             "version_added": "67"
           },
+          "deno": {
+            "version_added": "1.0",
+            "notes": "This interface is not exposed on the global scope on versions prior to 1.11."
```
Comment (ddbeck) [inline api/TransformStreamDefaultController.json:15]: Another likely partial_implementation

**Review by ddbeck — CHANGES_REQUESTED** [review]
Thank you for your patience while I reviewed this, @lucacasonato!

I've gotten through the `api/` section today. I'm trusting that your data is correct, but I looked for structural and style issues. I caught a few things of interest and I'll try to return to the javascript data soon (though feel free to make changes implied by my existing comments, if there's anything you'd already recognize).

File: api/CryptoKeyPair.json (hunk 1 above)
Comment (lucacasonato) [reply api/CryptoKeyPair.json:15]: Thanks - that makes sense. My intention for everywhere I added `"version_added": false` was to convey the message that it is not _yet_ implemented (ie we do intend to implement it, and there are open issues). For APIs that we are not planning to add as of now (DOM or Service Worker) I simply didn't add a `"deno"` key to the file.

File: api/Request.json (hunk 6 above)
Comment (lucacasonato) [reply api/Request.json]: Rephrased, and added to `Request` constructor and `WindowOrWorkerGlobalScope.fetch`.

File: api/TextDecoder.json (hunk 7 above)
Comment (lucacasonato) [reply api/TextDecoder.json]: Have clarified what "not supported" means and have updated the style. I can revisit this with the "proper" fix in a follow-up PR.

File: api/DedicatedWorkerGlobalScope.json [hunk 9]
```diff
@@ -359,6 +380,34 @@
...
+                  "The <code>transfer</code> argument is ignored."
+                ]
+              },
+              {
+                "version_added": "1.0",
+                "version_removed": "1.10",
+                "partial_implementation": true,
+                "notes": [
+                  "Data passed in the <code>message</code> property is serialized with JSON, not the structured clone algorithm.",
+                  "The <code>transfer</code> argument is ignored."
```
Comment (ddbeck) [inline api/DedicatedWorkerGlobalScope.json:407]: For consistency with MDN and some of our data guidelines, I'm going to suggest the word `parameter`

```suggestion
                  "The <code>message</code> parameter does not support cloning <code>SharedArrayBuffer</code> or <code>Blob</code> values.",
                  "The <code>transfer</code> parameter does not accept <code>ArrayBuffer</code> items. Passing an <code>ArrayBuffer</code> results in an error being thrown."
                ]
              },
              {
                "version_added": "1.10",
                "version_removed": "1.11",
                "partial_implementation": true,
                "notes": [
                  "The <code>message</code> parameter does not support <code>SharedArrayBuffer</code>.",
                  "The <code>transfer</code> parameter is ignored."
                ]
              },
              {
                "version_added": "1.0",
                "version_removed": "1.10",
                "partial_implementation": true,
                "notes": [
                  "Data passed in the <code>message</code> parameter is serialized with JSON, not the structured clone algorithm.",
                  "The <code>transfer</code> parameter is ignored."
```

File: api/MessageEvent.json [hunk 10]
```diff
@@ -308,6 +326,9 @@
               "chrome_android": {
                 "version_added": true
               },
+              "deno": {
+                "version_added": null
```
Comment (ddbeck) [inline api/MessageEvent.json:330]: Since it's pretty rare, I'm OK with this being `null` for now, but we are aiming for 100% "real" values in the API tree (i.e., actual version numbers and not `null`/`true`/`false` values). It would be nice to see this become a real value before merging, or in a follow-up PR.

See https://github.com/mdn/browser-compat-data/issues/6369 for more about the goal.

File: api/MessagePort.json [hunk 11]
```diff
@@ -333,6 +351,14 @@
             "chrome_android": {
               "version_added": "18"
             },
+            "deno": {
+              "version_added": "1.12",
+              "partial_implementation": true,
+              "notes": [
+                "The <code>message</code> property does not support cloning <code>SharedArrayBuffer</code> or <code>Blob</code> values.",
+                "The <code>transfer</code> argument does not accept <code>ArrayBuffer</code> items. Passing an <code>ArrayBuffer</code> results in an error being thrown."
```
Comment (ddbeck) [inline api/MessagePort.json:359]:
```suggestion
                "The <code>message</code> parameter does not support cloning <code>SharedArrayBuffer</code> or <code>Blob</code> values.",
                "The <code>transfer</code> parameter does not accept <code>ArrayBuffer</code> items. Passing an <code>ArrayBuffer</code> results in an error being thrown."
```

File: api/Request.json [hunk 12]
```diff
@@ -11,6 +11,10 @@
           "chrome_android": {
             "version_added": "42"
           },
+          "deno": {
+            "version_added": "1.0",
+            "notes": "Fetching with a <code>\"manual\"</code> redirect mode does not result in a <code>\"opaqueredirect\"</code> response, but a regular response."
```
Comment (ddbeck) [inline api/Request.json:16]: Perhaps this note could go on the constructor (the next feature down), instead?

File: api/Worker.json [hunk 13]
```diff
@@ -629,6 +660,34 @@
...
+                ]
+              },
+              {
+                "version_added": "1.0",
+                "version_removed": "1.10",
+                "partial_implementation": true,
+                "notes": [
+                  "Data passed in the <code>message</code> property is serialized with JSON, not the structured clone algorithm.",
+                  "The <code>transfer</code> argument is not supported, and results in an error being thrown."
+                ]
```
Comment (ddbeck) [inline api/Worker.json:688]:
```suggestion
                "notes": [
                  "The <code>message</code> parameter does not support cloning <code>SharedArrayBuffer</code> or <code>Blob</code> values.",
                  "The <code>transfer</code> parameter does not accept <code>ArrayBuffer</code> items. Passing an <code>ArrayBuffer</code> results in an error being thrown."
                ]
              },
              {
                "version_added": "1.10",
                "version_removed": "1.11",
                "partial_implementation": true,
                "notes": [
                  "The <code>message</code> parameter does not support <code>SharedArrayBuffer</code>.",
                  "The <code>transfer</code> parameter is not supported, and results in an error being thrown."
                ]
              },
              {
                "version_added": "1.0",
                "version_removed": "1.10",
                "partial_implementation": true,
                "notes": [
                  "Data passed in the <code>message</code> parameter is serialized with JSON, not the structured clone algorithm.",
                  "The <code>transfer</code> parameter is not supported, and results in an error being thrown."
                ]
```

File: javascript/builtins/Array.json [hunk 14]
```diff
@@ -1929,6 +2037,16 @@
...
                   "version_added": "25"
                 },
+                "deno": [
+                  {
+                    "version_added": "1.8"
+                  },
+                  {
+                    "version_added": "1.0",
+                    "partial_implementation": true,
+                    "notes": "Before version 1.8, only the locale data for <code>en-US</code> is available."
```
Comment (ddbeck) [inline javascript/builtins/Array.json:2047]: There's a weird subtlety to keep in mind when there are multiple support statements: if two support statements overlap, they need to make sense together. As written, I'd read it as:

- From version 1.8, supported with no limitations.
- From version 1.0, partially supported (with a limitation, noted).

These don't really make sense together (they could if, for example, one had an alternative name and the other did not). I suspect this is what you really meant:

- From version 1.8, supported with no limitations.
- From version 1.0 to 1.8, partially supported (with a limitation, noted).

And the source changes to make that so:

```suggestion
                    "version_added": "1.0",
                    "version_removed": "1.8"
                    "partial_implementation": true,
                    "notes": "Only the locale data for <code>en-US</code> is available."
```

File: javascript/builtins/BigInt.json [hunk 15]
```diff
@@ -269,6 +284,16 @@
...
                   "version_added": "76"
                 },
+                "deno": [
+                  {
+                    "version_added": "1.8"
+                  },
+                  {
+                    "version_added": "1.0",
+                    "partial_implementation": true,
+                    "notes": "Before version 1.8, only the locale data for <code>en-US</code> is available."
```
Comment (ddbeck) [inline javascript/builtins/BigInt.json:294]:
```suggestion
                    "version_added": "1.0",
                    "version_removed": "1.8"
                    "partial_implementation": true,
                    "notes": "Only the locale data for <code>en-US</code> is available."
```

File: javascript/builtins/Date.json [hunk 16]
```diff
@@ -2455,6 +2596,16 @@
...
                   "version_added": "25"
                 },
+                "deno": [
+                  {
+                    "version_added": "1.8"
+                  },
+                  {
+                    "version_added": "1.0",
+                    "partial_implementation": true,
+                    "notes": "Before version 1.8, only the locale data for <code>en-US</code> is available."
```
Comment (ddbeck) [inline javascript/builtins/Date.json:2606]:
```suggestion
                    "version_added": "1.0",
                    "version_removed": "1.8",
                    "partial_implementation": true,
                    "notes": "Only the locale data for <code>en-US</code> is available."
```

File: javascript/builtins/Date.json [hunk 17]
```diff
@@ -2668,6 +2828,16 @@
...
+                  }
```
Comment (ddbeck) [inline javascript/builtins/Date.json:2839]:
```suggestion
                  {
                    "version_added": "1.0",
                    "version_removed": "1.8",
                    "partial_implementation": true,
                    "notes": "Only the locale data for <code>en-US</code> is available."
                  }
```

File: javascript/builtins/Date.json [hunk 18]
```diff
@@ -2881,6 +3060,16 @@
...
```
Comment (ddbeck) [inline javascript/builtins/Date.json:3071]:
```suggestion
                  {
                    "version_added": "1.0",
                    "version_removed": "1.8",
                    "partial_implementation": true,
                    "notes": "Only the locale data for <code>en-US</code> is available."
                  }
```

File: javascript/builtins/Number.json [hunk 19]
```diff
@@ -1002,6 +1059,16 @@
...
                 },
+                "deno": [
+                  {
+                    "version_added": "1.8"
+                  },
+                  {
+                    "version_added": "1.0",
+                    "partial_implementation": true,
+                    "notes": "Before version 1.8, only the locale data for <code>en-US</code> is available."
+                  }
```
Comment (ddbeck) [inline javascript/builtins/Number.json:1070]:
```suggestion
                  {
                    "version_added": "1.0",
                    "version_removed": "1.8",
                    "partial_implementation": true,
                    "notes": "Only the locale data for <code>en-US</code> is available."
                  }
```

**Review by ddbeck — CHANGES_REQUESTED** [review]
All right, we're chipping away things here! I've reviewed through the first 100 files.

File: api/MessageEvent.json (hunk 10 above)
Comment (lucacasonato) [reply api/MessageEvent.json]: Updated it to `false` (checked implementation).

File: api/Request.json (hunk 12 above)
Comment (lucacasonato) [reply api/Request.json]: Done

File: javascript/builtins/Array.json [hunk 20]
```diff
@@ -1929,6 +2037,17 @@
                 "chrome_android": {
                   "version_added": "25"
                 },
+                "deno": [
+                  {
+                    "version_added": "1.8"
+                  },
+                  {
+                    "version_added": "1.0",
+                    "version_removed": "1.8"
```
Comment (ddbeck) [inline javascript/builtins/Array.json:2046]:
```suggestion
                    "version_removed": "1.8",
```

File: javascript/builtins/BigInt.json [hunk 21]
```diff
@@ -269,6 +284,17 @@
                 "chrome_android": {
                   "version_added": "76"
                 },
+                "deno": [
+                  {
+                    "version_added": "1.8"
+                  },
+                  {
+                    "version_added": "1.0",
+                    "version_removed": "1.8"
```
Comment (ddbeck) [inline javascript/builtins/BigInt.json:293]: This is my fault. Sorry!

```suggestion
                    "version_removed": "1.8",
```

File: javascript/builtins/String.json [hunk 22]
```diff
@@ -1308,6 +1380,16 @@
...
                 },
+                "deno": [
+                  {
+                    "version_added": "1.8"
+                  },
+                  {
+                    "version_added": "1.0",
+                    "partial_implementation": true,
+                    "notes": "Before version 1.8, only the locale data for <code>en-US</code> is available."
+                  }
```
Comment (ddbeck) [inline javascript/builtins/String.json:1391]:
```suggestion
                  {
                    "version_added": "1.0",
                    "version_removed": "1.8",
                    "partial_implementation": true,
                    "notes": "Only the locale data for <code>en-US</code> is available."
                  }
```

File: javascript/builtins/String.json [hunk 23]
```diff
@@ -2562,6 +2710,16 @@
...
```
Comment (ddbeck) [inline javascript/builtins/String.json:2721]:
```suggestion
                  {
                    "version_added": "1.0",
                    "version_removed": "1.8",
                    "partial_implementation": true,
                    "notes": "Only the locale data for <code>en-US</code> is available."
                  }
```

File: javascript/builtins/String.json [hunk 24]
```diff
@@ -2674,6 +2835,16 @@
...
```
Comment (ddbeck) [inline javascript/builtins/String.json:2846]:
```suggestion
                  {
                    "version_added": "1.0",
                    "version_removed": "1.8",
                    "partial_implementation": true,
                    "notes": "Only the locale data for <code>en-US</code> is available."
                  }
```

File: schemas/compat-data-schema.md [hunk 25]
```diff
@@ -120,6 +120,7 @@ The currently accepted browser identifiers should be declared in alphabetical or
 
 - `chrome`, Google Chrome (on desktops)
 - `chrome_android`, Google Chrome (on Android)
+- `deno` Deno JavaScript runtime built on Chrome's V8 JavaScript engine
```
Comment (ddbeck) [inline schemas/compat-data-schema.md:123]: tiniest consistency fix

```suggestion
- `deno`, Deno JavaScript runtime built on Chrome's V8 JavaScript engine
```

**Review by ddbeck — CHANGES_REQUESTED** [review]
A few more suggestions for you. We're getting really close now!

**Comment by ddbeck** [comment]
The latest changes look good! Looks like a merge or rebase is needed to resolve a new conflict, though.

**Comment by lucacasonato** [comment]
@ddbeck Merged with main!

File: browsers/deno.json [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+  "browsers": {
+    "deno": {
+      "name": "Deno",
+      "releases": {
+        "1.0": {
+          "release_date": "2020-05-13",
+          "release_notes": "https://github.com/denoland/deno/releases/tag/v1.0.0",
+          "status": "retired",
+          "engine": "V8",
+          "engine_version": "8.4"
```
Comment (zloirock) [inline browsers/deno.json:11]: Where could I find an official and parsable source of information about matching V8 versions to Deno versions, like https://nodejs.org/dist/index.json ?

Comment (lucacasonato) [reply browsers/deno.json:11]: We do not have that information in programmatic form right now, but I have opened https://github.com/denoland/deno/issues/11195 to track creation of such a file. If you need info about a specific file, V8 versions are printed when running `deno --version`, and are available at `Deno.version.v8` at runtime. You can also upgrade and downgrade to any version using `deno upgrade --version=$version`, and you can find a list of all versions at https://raw.githubusercontent.com/denoland/deno_website2/main/versions.json.

**Comment by ddbeck** [comment]
@lucacasonato a brief update here: I wasn't quite ready to merge this for yesterday's release (I need to do a little more prep for release notes, etc., which was hard to do while my colleagues at Mozilla were away) and I'll be off next week. So I'll come back to this the week of July 12. My apologies for putting this on hold—poor timing with folks being out at different times.

**Comment by lucacasonato** [comment]
No problem. Have a nice week off! 🏖

**Review by ddbeck — CHANGES_REQUESTED** [review]
Two tiny things to address. Thank you!

File: api/SubtleCrypto.json [hunk 27]
```diff
@@ -12,7 +12,8 @@
             "version_added": "37"
           },
           "deno": {
-            "version_added": "1.11"
+            "version_added": "1.11",
+            "partial_implementation": true
```
[the remark this answers is not in the document]
Comment (lucacasonato) [reply api/SubtleCrypto.json]: I had set it to partial because some of the descendants are `"version_added": false`. That doesn't make much sense though as you said, so I have removed the `"partial_implementation"` from the interface. It is still present on some descendants, but with `"notes"`.

File: browsers/deno.json [hunk 28]
```diff
@@ -83,16 +83,30 @@
...
           "release_date": "2020-07-13",
           "release_notes": "https://github.com/denoland/deno/releases/tag/v1.12.0",
-          "status": "nightly",
+          "status": "current",
           "engine": "V8",
           "engine_version": "9.2"
+        },
+        "1.13": {
+          "release_date": "2020-08-10",
+          "release_notes": "https://github.com/denoland/deno/releases/tag/v1.13.0",
```
[the remark this answers is not in the document]
Comment (lucacasonato) [reply browsers/deno.json:99]: Yup, it will start working tomorrow. I have removed the url from 1.14 though, as that is still a few weeks out.

**Review by ddbeck — APPROVED** [review]
Thank you so much for your care and patience with this, @lucacasonato. I'm pleased to approve this and I'll be merging momentarily. These changes will be included in v4.0.0, which I aim to ship later in the week.

With that, let me be the first to say: welcome to BCD, Deno! 🎉

**Comment by lucacasonato** [comment]
Thanks Daniel!
